buildkite/resource_agent_token.go

```diff
@@ -48,82 +46,74 @@ func resourceAgentToken() *schema.Resource {
 
 // CreateToken creates a Buildkite agent token
 func CreateToken(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
 	client := m.(*Client)
 
-	var mutation struct {
-		AgentTokenCreate struct {
-			AgentTokenEdge struct {
-				Node AgentTokenNode
-			}
-		} `graphql:"agentTokenCreate(input: {organizationID: $org, description: $desc})"`
-	}
-
-	vars := map[string]interface{}{
-		"org":  client.organizationId,
-		"desc": graphql.String(d.Get("description").(string)),
-	}
+	apiResponse, err := createAgentToken(
+		client.genqlient,
+		client.organizationId,
+		d.Get("description").(string),
+	)
 
-	err := client.graphql.Mutate(context.Background(), &mutation, vars)
 	if err != nil {
 		return diag.FromErr(err)
 	}
 
-	updateAgentToken(d, &mutation.AgentTokenCreate.AgentTokenEdge.Node)
+	d.SetId(apiResponse.AgentTokenCreate.AgentTokenEdge.Node.Id)
+	d.Set("uuid", apiResponse.AgentTokenCreate.AgentTokenEdge.Node.Uuid)
+	d.Set("description", apiResponse.AgentTokenCreate.AgentTokenEdge.Node.Description)
+	d.Set("token", apiResponse.AgentTokenCreate.TokenValue)
 
-	return nil
+	return diags
 }
 
 // ReadToken retrieves a Buildkite agent token
 func ReadToken(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
 	client := m.(*Client)
-	var query struct {
-		Node struct {
-			AgentToken AgentTokenNode `graphql:"... on AgentToken"`
-		} `graphql:"node(id: $id)"`
-	}
 
-	vars := map[string]interface{}{
-		"id": d.Id(),
-	}
+	agentToken, err := getAgentToken(client.genqlient, fmt.Sprintf("%s/%s", client.organization, d.Get("uuid").(string)))
 
-	err := client.graphql.Query(context.Background(), &query, vars)
 	if err != nil {
 		return diag.FromErr(err)
 	}
-	if query.Node.AgentToken.RevokedAt != "" {
-		return diag.FromErr(errors.New("Cannot import revoked token"))
+
+	if agentToken.AgentToken.Id == "" {
+		return diag.FromErr(errors.New("Agent Token not found"))
 	}
 
-	updateAgentToken(d, &query.Node.AgentToken)
+	d.SetId(agentToken.AgentToken.Id)
+	d.Set("uuid", agentToken.AgentToken.Uuid)
+	d.Set("description", agentToken.AgentToken.Description)
+	// NB: we never set the token in read context because its not available in the API after creation
 
-	return nil
+	return diags
 }
 
-// DeleteToken revokes a Buildkite agent token - they cannot be completely deleted
+// DeleteToken revokes a Buildkite agent token - they cannot be completely deleted (will have a revoke)
 func DeleteToken(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
 	client := m.(*Client)
-	var mutation struct {
-		AgentTokenRevoke struct {
-			AgentToken AgentTokenNode
-		} `graphql:"agentTokenRevoke(input: {id: $id, reason: $reason})"`
+
+	agentTokenRevCandidate, err := getAgentToken(client.genqlient, fmt.Sprintf("%s/%s", client.organization, d.Get("uuid").(string)))
```

Comment:
I'm curious why you are getting the token here first. Can you explain the reasoning?

Reply:
With the delete fn - its more I think so that agentTokenRevCandidate can be had without the contextual d.Id()

Reply:
Oh right okay. No the `d.Id()` will be in state so that can be used directly and we save on an extra API request. So I think you can remove a few lines here and jump straight to `revokeAgentToken()`